kraken/lib/pretrain/layers.py

```diff
@@ -0,0 +1,126 @@
+"""
+Layers for VGSL models
+"""
+import torch
+
+from typing import Tuple, Optional
+from torch.nn import Module, Embedding, Linear
+from torch.nn.utils.rnn import pad_packed_sequence, pack_padded_sequence
+
+from kraken.lib.vgsl import VGSLBlock
+from kraken.lib.pretrain.util import compute_masks
+
+# all tensors are ordered NCHW, the "feature" dimension is C, so the output of
+# an LSTM will be put into C same as the filters of a CNN.
+
+__all__ = ['Wav2Vec2Mask']
+
+
+class Wav2Vec2Mask(Module):
+    """
+    A layer for Wav2Vec2-style masking. Needs to be placed just before
+    recurrent/contextual layers.
+    """
+
+    def __init__(self,
+                 context_encoder_input_dim: int,
+                 final_dim: int,
+                 mask_width: int,
+                 mask_prob: float,
+                 num_negatives: int) -> None:
+        """
+
+        Args:
+            context_encoder_input_dim: size of the `C` input dimension
+            final_dim: size of the decoder `C` output dimension just before the
+                       final linear projection.
+            mask_width: width of the non-overlapping masked areas.
+            mask_prob: probability of masking at each time step
+            num_negatives: number of negative samples with width mask_width *
+                           num_masks
+
+        Shape:
+            - Inputs: :math:`(N, C, H, W)` where `N` batches, `C` channels, `H`
+              height, and `W` width.
+            - Outputs output :math:`(N, C, H, W)`
+        """
+        super().__init__()
+
+        self.context_encoder_input_dim = context_encoder_input_dim
+        self.final_dim = final_dim
+        self.mask_width = mask_width
+        self.mask_prob = mask_prob
+        self.num_negatives = num_negatives
+
+        # mask embedding replacing the masked out areas
+        self.mask_emb = Embedding(mask_width, context_encoder_input_dim)
```

Comment:
it's maybe worth having a _separate_ knob we can set for width of the mask embedding itself, i.e., so that we can try using either a `mask_width` embedding here or a 1 time-step wide mask that gets repeated like original paper

Reply:
I'm loath to add switches where it should be possible to find an 'optimal' value. I'd evaluate both this formulation and the original one and see which one performs better and then just use that one.